File: src/main/java/uno_ui/NetworkClient.java
package uno_ui;

import uno_proto.common.*;
import uno_proto.dto.*;
import uno_server.protocol.MessageParser;

import java.io.*;
import java.net.Socket;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.function.Consumer;

/**
 * NetworkClient handles all communication with the UNO server.
 * It runs on a background thread to avoid blocking the UI.
 * 
 * This class:
 * - Connects to the server via TCP socket
 * - Sends NetworkMessage objects as JSON
 * - Receives and parses JSON responses
 * - Notifies listeners when messages arrive
 */
public class NetworkClient {
    
    // Server connection details
    private static final String HOST = "localhost";
    private static final int PORT = 9090;
    
    // Socket and I/O streams for communicating with server
    private Socket socket;
    private PrintWriter out;
    private BufferedReader in;
    
    // Parser for converting between JSON and NetworkMessage objects
    private final MessageParser parser;
    
    // Message counter for giving each message a unique ID
    private long messageIdCounter = 1;
    
    // Queue for messages we want to send (thread-safe)
    private final BlockingQueue<NetworkMessage> outgoingMessages;
    
    // Callback that gets called when we receive a message from server
    private Consumer<NetworkMessage> messageListener;
    
    // Background threads for sending and receiving
    private Thread senderThread;
    private Thread receiverThread;
    
    // Flag to control when threads should stop
    private volatile boolean running = false;
    
    /**
     * Creates a new NetworkClient.
     * Note: This doesn't connect yet - call connect() to start.
     */
    public NetworkClient() {
        this.parser = new MessageParser();
        this.outgoingMessages = new LinkedBlockingQueue<>();
    }
    
    /**
     * Sets the callback that will be invoked when messages arrive from server.
     * This callback will be called from the receiver thread, so if you need to
     * update UI, make sure to use Platform.runLater()!
     * 
     * @param listener Function that receives NetworkMessage objects
     */
    public void setMessageListener(Consumer<NetworkMessage> listener) {
        this.messageListener = listener;
    }
    
    /**
     * Connects to the server and starts background threads.
     * 
     * @return true if connection successful, false otherwise
     */
    public boolean connect() {
        try {
            // Try to connect to server
            System.out.println("[NetworkClient] Connecting to " + HOST + ":" + PORT + "...");
            socket = new Socket(HOST, PORT);
            out = new PrintWriter(socket.getOutputStream(), true);
            in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            
            System.out.println("[NetworkClient] Connected!");
            
            // Start background threads
            running = true;
            startSenderThread();
            startReceiverThread();
            
            return true;
        } catch (IOException e) {
            System.err.println("[NetworkClient] Failed to connect: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Disconnects from server and stops all threads.
     */
    public void disconnect() {
        System.out.println("[NetworkClient] Disconnecting...");
        running = false;
        
        // Close socket (this will also cause threads to exit)
        try {
            if (socket != null && !socket.isClosed()) {
                socket.close();
            }
        } catch (IOException e) {
            System.err.println("[NetworkClient] Error closing socket: " + e.getMessage());
        }
        
        // Wait for threads to finish
        try {
            if (senderThread != null) {
                senderThread.join(1000);
            }
            if (receiverThread != null) {
                receiverThread.join(1000);
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        System.out.println("[NetworkClient] Disconnected");
    }
    
    /**
     * Sends a message to the server.
     * The message is queued and will be sent by the sender thread.
     * 
     * @param method The protocol method to invoke
     * @param payload The data to send (use MessageParser.EmptyPayload() if none)
     */
    public void sendMessage(Method method, Payload payload) {
        // Create a NetworkMessage with auto-incrementing ID
        NetworkMessage message = new NetworkMessage(
            messageIdCounter++,
            Version.V1,
            method,
            payload,
            System.currentTimeMillis()
        );
        
        // Add to queue - sender thread will pick it up
        outgoingMessages.offer(message);
    }
    
    /**
     * Starts the sender thread which sends queued messages to server.
     */
    private void startSenderThread() {
        senderThread = new Thread(() -> {
            System.out.println("[Sender] Thread started");
            
            while (running) {
                try {
                    // Wait for a message to send (blocks until one is available)
                    // Timeout after 1 second to check if we should still be running
                    NetworkMessage message = outgoingMessages.poll(1, java.util.concurrent.TimeUnit.SECONDS);
                    
                    if (message != null) {
                        // Convert to JSON and send
                        String json = parser.toJson(message);
                        System.out.println("[Sender] Sending: " + json);
                        out.println(json);
                    }
                } catch (InterruptedException e) {
                    // Thread interrupted - exit loop
                    break;
                } catch (Exception e) {
                    System.err.println("[Sender] Error sending message: " + e.getMessage());
                }
            }
            
            System.out.println("[Sender] Thread stopped");
        }, "NetworkClient-Sender");
        
        senderThread.setDaemon(true);
        senderThread.start();
    }
    
    /**
     * Starts the receiver thread which reads messages from server.
     */
    private void startReceiverThread() {
        receiverThread = new Thread(() -> {
            System.out.println("[Receiver] Thread started");
            
            while (running) {
                try {
                    // Read one line of JSON from server
                    String line = in.readLine();
                    
                    if (line == null) {
                        // Server closed connection
                        System.out.println("[Receiver] Server closed connection");
                        break;
                    }
                    
                    System.out.println("[Receiver] Received: " + line);
                    
                    // Parse JSON into NetworkMessage
                    NetworkMessage message = parser.fromJson(line);
                    
                    // Notify listener if one is registered
                    if (messageListener != null) {
                        messageListener.accept(message);
                    }
                    
                } catch (IOException e) {
                    if (running) {
                        System.err.println("[Receiver] Error reading from server: " + e.getMessage());
                    }
                    break;
                } catch (Exception e) {
                    System.err.println("[Receiver] Error parsing message: " + e.getMessage());
                    e.printStackTrace();
                }
            }
            
            System.out.println("[Receiver] Thread stopped");
        }, "NetworkClient-Receiver");
        
        receiverThread.setDaemon(true);
        receiverThread.start();
    }
    
    /**
     * Checks if we're currently connected to server.
     * 
     * @return true if connected, false otherwise
     */
    public boolean isConnected() {
        return socket != null && !socket.isClosed() && running;
    }
}

File: src/main/java/uno_ui/NetworkClient.kt
package uno_ui

import uno_proto.common.*
import uno_server.protocol.MessageParser
import java.io.*
import java.net.Socket
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

/**
 * Handles TCP socket communication with the server.
 * Sends and receives NetworkMessage objects.
 * Includes heartbeat mechanism to keep connection alive.
 */
class NetworkClient(
    private val host: String = "localhost",
    private val port: Int = 9090
) {
    private var socket: Socket? = null
    private var out: PrintWriter? = null
    private var input: BufferedReader? = null

    private val serializer = MessageSerializer()
    private val outgoingMessages: BlockingQueue<NetworkMessage> = LinkedBlockingQueue()
    
    private var messageListener: Consumer<NetworkMessage>? = null
    private var senderThread: Thread? = null
    private var receiverThread: Thread? = null
    private var heartbeatThread: Thread? = null
    
    @Volatile
    private var running = false
    
    @Volatile
    private var lastPongTime = AtomicLong(System.currentTimeMillis())
    
    @Volatile
    private var lastPingSentTime = AtomicLong(0L)
    
    private var messageIdCounter = 1L
    
    companion object {
        private const val HEARTBEAT_INTERVAL_MS = 30000L // 30 seconds
        private const val PONG_TIMEOUT_MS = 10000L // 10 seconds
    }

    fun setMessageListener(listener: (NetworkMessage) -> Unit) {
        messageListener = Consumer(listener)
    }

    fun connect(): Boolean {
        return try {
            println("[NetworkClient] Connecting to $host:$port...")
            socket = Socket(host, port)
            out = PrintWriter(socket!!.getOutputStream(), true)
            input = BufferedReader(InputStreamReader(socket!!.getInputStream()))
            
            println("[NetworkClient] Connected!")
            
            running = true
            lastPongTime.set(System.currentTimeMillis())
            lastPingSentTime.set(0L)
            
            startSenderThread()
            startReceiverThread()
            startHeartbeatThread()
            
            true
        } catch (e: IOException) {
            System.err.println("[NetworkClient] Failed to connect: ${e.message}")
            false
        }
    }

    fun disconnect() {
        println("[NetworkClient] Disconnecting...")
        running = false
        
        try {
            socket?.takeIf { !it.isClosed }?.close()
        } catch (e: IOException) {
            System.err.println("[NetworkClient] Error closing socket: ${e.message}")
        }
        
        senderThread?.let {
            it.join(1000)
        }
        receiverThread?.let {
            it.join(1000)
        }
        heartbeatThread?.let {
            it.join(1000)
        }
        
        println("[NetworkClient] Disconnected")
    }

    fun sendMessage(method: Method, payload: Payload) {
        val message = NetworkMessage(
            messageIdCounter++,
            Version.V1,
            method,
            payload,
            System.currentTimeMillis()
        )
        outgoingMessages.offer(message)
    }

    private fun startSenderThread() {
        senderThread = Thread({
            println("[Sender] Thread started")
            
            while (running) {
                try {
                    val message = outgoingMessages.poll(1, TimeUnit.SECONDS)
                    
                    if (message != null) {
                        val json = serializer.serialize(message)
                        println("[Sender] Sending: $json")
                        out?.println(json)
                    }
                } catch (e: InterruptedException) {
                    break
                } catch (e: Exception) {
                    System.err.println("[Sender] Error sending message: ${e.message}")
                }
            }
            
            println("[Sender] Thread stopped")
        }, "NetworkClient-Sender").apply {
            isDaemon = true
            start()
        }
    }

    private fun startReceiverThread() {
        receiverThread = Thread({
            println("[Receiver] Thread started")
            
            while (running) {
                try {
                    val line = input?.readLine()
                    
                    if (line == null) {
                        println("[Receiver] Server closed connection")
                        break
                    }
                    
                    println("[Receiver] Received: $line")
                    
                    val message = serializer.deserialize(line)
                    message?.let { 
                        // Handle PONG internally for heartbeat
                        if (it.method == Method.PONG) {
                            lastPongTime.set(System.currentTimeMillis())
                            println("[Heartbeat] PONG received")
                        }
                        // Notify listener for all messages (including PONG if they want to handle it)
                        messageListener?.accept(it) 
                    }
                    
                } catch (e: IOException) {
                    if (running) {
                        System.err.println("[Receiver] Error reading from server: ${e.message}")
                    }
                    break
                } catch (e: Exception) {
                    System.err.println("[Receiver] Error parsing message: ${e.message}")
                    e.printStackTrace()
                }
            }
            
            println("[Receiver] Thread stopped")
        }, "NetworkClient-Receiver").apply {
            isDaemon = true
            start()
        }
    }

    private fun startHeartbeatThread() {
        heartbeatThread = Thread({
            println("[Heartbeat] Thread started")
            
            while (running) {
                try {
                    Thread.sleep(HEARTBEAT_INTERVAL_MS)
                    
                    if (!running) break
                    
                    // Send PING
                    println("[Heartbeat] Sending PING...")
                    lastPingSentTime.set(System.currentTimeMillis())
                    sendMessage(Method.PING, MessageParser.EmptyPayload)
                    
                    // Wait a bit to allow PONG to arrive
                    Thread.sleep(PONG_TIMEOUT_MS)
                    
                    if (!running) break
                    
                    // Check if we received PONG in time
                    val timeSinceLastPong = System.currentTimeMillis() - lastPongTime.get()
                    val timeSincePing = System.currentTimeMillis() - lastPingSentTime.get()
                    
                    if (timeSincePing < PONG_TIMEOUT_MS + 1000) {
                        // We just sent a ping, check if we got a pong
                        if (timeSinceLastPong > HEARTBEAT_INTERVAL_MS + PONG_TIMEOUT_MS) {
                            System.err.println("[Heartbeat] No PONG received in ${PONG_TIMEOUT_MS}ms, reconnecting...")
                            reconnect()
                            break
                        }
                    }
                    
                } catch (e: InterruptedException) {
                    break
                } catch (e: Exception) {
                    System.err.println("[Heartbeat] Error: ${e.message}")
                    e.printStackTrace()
                }
            }
            
            println("[Heartbeat] Thread stopped")
        }, "NetworkClient-Heartbeat").apply {
            isDaemon = true
            start()
        }
    }
    
    private fun reconnect() {
        println("[NetworkClient] Attempting to reconnect...")
        disconnect()
        Thread.sleep(1000) // Wait a bit before reconnecting
        connect()
    }

    fun isConnected(): Boolean = socket?.let { !it.isClosed && running } ?: false
}
